src/classes/adapter.js

```diff
@@ -0,0 +1,144 @@
+   * @constructor
+   */
+  constructor (config) {
+    super()
+    this.config = config
+    this.knex = Knex({
+      client: 'pg',
+      connection: config,
+      searchPath: 'knex,public'
+    })
+  }
+
+  /**
+   * Converts a models attributes hash into an array of column names for use in
+   * an sql query
+   *
+   * @param {Object} attributes - attributes hash of the form
+   * @example {title: 'string', createdAt: 'date'}
+   * @return {Array} - an array of column names
+   */
+  fieldsFromAttributes (attributes) {
+    return Object.keys(attributes)
+  }
+
+  parseIncludes () {
+
+  }
+
+  include () {
+
+  }
+
+  /**
+   * Finds all records for a given model. @see `query` when you need more control over
+   * what records are returned.
+   *
+   * @param {Object} Model - object with properties `tableName` and an `attributes` hash
+   * @return {Array} - array of database records
+   */
+  findAll (Model) {
+    const attributes = this.fieldsFromAttributes(Model.definition.attributes)
+    const tableName = Model.tableName
+    return this.knex.column(attributes).select().from(tableName)
+  }
+
+  /*
+  options should be implemented here to jsonapi spec
+  {
+    include: 'comments, posts, comments.author',
+    fields: {
+      articles: 'title, body',
+      people: 'name'
+    },
+    sort: 'age, -name',
+    page: {number: 1, size: 20},
+    filter: {}
+  }
+  */
+  query (Model, options) {
+    const attributes = this.fieldsFromAttributes(Model.definition.attributes)
+    const tableName = Model.tableName
+
+    if (typeof options.fields === 'object') {
+      const types = Object.keys(options.fields)
+      if (types.indexOf(Model.type) !== -1) {
+        const fields = options.fields[Model.type].split(',').map(field => field.trim())
+        for (const attr of attributes) {
+          if (fields.indexOf(attr) === -1) {
+            attributes.splice(attributes.indexOf(attr))
+          }
+        }
+        attributes.push('id')
+      }
+    }
+
+    let query = this.knex.column(attributes).select().from(tableName)
+
+    if (typeof options.page === 'object' && options.page.number && options.page.size) {
+      query = query
+        .limit(options.page.size)
+        .offset(options.page.number * options.page.size - options.page.size)
+    }
+
+    if (typeof options.sort === 'string') {
+      options.sort.split(',').forEach(column => {
+        const direction = (column.indexOf('-') !== -1) ? 'DESC' : 'ASC'
+        column = column.replace('-', '')
+        query = query.orderBy(column.trim(), direction)
+      })
+    }
+
+    if (typeof options.include === 'string') {
+      const includes = this.parseIncludes(options.include)
+      for (const include of includes) {
+        // needs to recurse since we may need to get comments for posts but also
+        // author for comments
+        // I think the include is going to need to perform queries, create
+        // lists and then merge them
```

Comment:
ooo I always hate this about ORMs - because they are general and don't know anything about your data structure, they are really inefficient and usually cause exponential growth in the number of round trips to the DB every time you add an include.

10/10 if you're doing to stub out a feature like this, have it raise an error saying `new Error('Not Implemented: This feature not supported yet')` instead of just silently doing nothing.